server/src/main/java/org/elasticsearch/transport/InboundPipeline.java

```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
     public void doHandleBytes(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
         channel.getChannelStats().markAccessed(relativeTimeInMillis.getAsLong());
         statsTracker.markBytesRead(reference.length());
-        pending.add(reference.retain());
 
-        final ArrayList<Object> fragments = fragmentList.get();
-        boolean continueHandling = true;
-
-        while (continueHandling && isClosed == false) {
-            boolean continueDecoding = true;
-            while (continueDecoding && pending.isEmpty() == false) {
-                try (ReleasableBytesReference toDecode = getPendingBytes()) {
-                    final int bytesDecoded = decoder.decode(toDecode, fragments::add);
-                    if (bytesDecoded != 0) {
-                        releasePendingBytes(bytesDecoded);
-                        if (fragments.isEmpty() == false && endOfMessage(fragments.get(fragments.size() - 1))) {
-                            continueDecoding = false;
-                        }
-                    } else {
-                        continueDecoding = false;
-                    }
-                }
-            }
+        if (pending.isEmpty() == false) {
+            // we already have pending bytes, so we queue these bytes after them and then try to decode from the combined message
+            pending.add(reference.retain());
+            doHandleBytesWithPending(channel);
+            return;
+        }
 
-            if (fragments.isEmpty()) {
-                continueHandling = false;
+        while (isClosed == false && reference.length() > 0) {
+            final int bytesDecoded = decode(channel, reference);
+            if (bytesDecoded != 0) {
+                reference = reference.releasableSlice(bytesDecoded);
             } else {
-                try {
-                    forwardFragments(channel, fragments);
-                } finally {
-                    for (Object fragment : fragments) {
-                        if (fragment instanceof ReleasableBytesReference) {
-                            ((ReleasableBytesReference) fragment).close();
-                        }
-                    }
-                    fragments.clear();
-                }
+                break;
             }
         }
+        // if handling the messages didn't cause the channel to get closed and we did not fully consume the buffer retain it
```

Comment:
Could we have a test in `InboundPipelineTests` showing that we handle the `isClosed` case correctly here? And elsewhere I guess, but I checked that we never actually exercise the `isClosed == true` branch here.

Reply:
Let me try :) Should be doable, these paths are exercised by some internal cluster tests so there should be an obvious way to do it. On it

Reply:
Test added, should be good for another round now :)